sklearn/semi_supervised/_self_training.py

```diff
@@ -0,0 +1,338 @@
+        -------
+        self : object
+            returns an instance of self.
+        """
+        X, y = check_X_y(X, y)
+
+        if self.base_estimator is None:
+            raise ValueError("base_estimator cannot be None!")
+
+        self.base_estimator_ = clone(self.base_estimator)
+
+        if self.max_iter is not None and self.max_iter < 0:
+            msg = "max_iter must be >= 0 or None, got {}".format(self.max_iter)
+            raise ValueError(msg)
+
+        if not 0 <= self.threshold < 1:
+            msg = "threshold must be in [0,1), got {}".format(self.threshold)
+            raise ValueError(msg)
+
+        if self.criterion not in ['threshold', 'k_best']:
+            raise ValueError("criterion must be either 'threshold' "
+                             "or 'k_best', got "
+                             "{}".format(self.criterion))
+
+        if y.dtype.kind in ['U', 'S']:
+            raise ValueError("y has dtype string. If you wish to predict on "
+                             "string targets, use dtype object, and use -1"
+                             " as the label for unlabeled samples.")
+
+        has_label = y != -1
+
+        if np.all(has_label):
+            warnings.warn("y contains no unlabeled samples", UserWarning)
+
+        if self.criterion == 'k_best' and (self.k_best > X.shape[0] -
+                                           np.sum(has_label)):
+            warnings.warn("k_best is larger than the amount of unlabeled "
+                          "samples. All unlabeled samples will be labeled in "
+                          "the first iteration", UserWarning)
+
+        self.transduction_ = np.copy(y)
+        self.labeled_iter_ = np.full_like(y, -1)
+        self.labeled_iter_[has_label] = 0
+
+        self.n_iter_ = 0
+
+        while not np.all(has_label) and (self.max_iter is None or
+                                         self.n_iter_ < self.max_iter):
+            self.n_iter_ += 1
+            self.base_estimator_.fit(
+                X[safe_mask(X, has_label)],
+                self.transduction_[safe_mask(self.transduction_, has_label)])
+
+            if self.n_iter_ == 1:
+                _validate_estimator(self.base_estimator)
+
+            # Predict on the unlabeled samples
+            prob = self.base_estimator_.predict_proba(
+                X[safe_mask(X, ~has_label)])
+            pred = self.base_estimator_.classes_[np.argmax(prob, axis=1)]
+            max_proba = np.max(prob, axis=1)
+
+            # Select samples
+            if self.criterion == 'threshold':
+                new_labels_unlabeled = max_proba > self.threshold
+            else:
+                n_to_select = min(self.k_best, max_proba.shape[0])
+                if n_to_select == max_proba.shape[0]:
+                    new_labels_unlabeled = np.ones_like(max_proba, dtype=bool)
+                else:
+                    # NB these are indicies, not a mask
+                    new_labels_unlabeled = \
```

Comment:
No, this doesn't work for both cases since argpartition throws a ValueError if k is out of bounds of the array (which is the case if we have to select all samples from max_proba). I guess you could do something like
```
selected =  np.argpartition(
    -max_proba, 
    min(n_to_select, max_proba.shape[0]-1)
)[:n_to_select]
```
but that seems less readable to me.